phoenix-core/src/main/java/org/apache/phoenix/execute/MutationState.java

```diff
@@ -931,220 +951,298 @@ private void send(Iterator<TableRef> tableRefIterator) throws SQLException {
-                            try {
-                                hTable.close();
-                            } catch (IOException e) {
-                                if (sqlE != null) {
-                                    sqlE.setNextException(ServerUtil.parseServerException(e));
-                                } else {
-                                    sqlE = ServerUtil.parseServerException(e);
-                                }
+                            hTable.close();
+                        } catch (IOException e) {
+                            if (sqlE != null) {
+                                sqlE.setNextException(ServerUtil.parseServerException(e));
+                            } else {
+                                sqlE = ServerUtil.parseServerException(e);
                             }
-                            if (sqlE != null) { throw sqlE; }
                         }
+                        if (sqlE != null) { throw sqlE; }
                     }
-                } while (shouldRetry && retryCount++ < 1);
+                }
+            } while (shouldRetry && retryCount++ < 1);
+        }
+    }
+
+    private void filterIndexCheckerMutations(Map<TableInfo, List<Mutation>> mutationMap,
+            Map<TableInfo, List<Mutation>> unverifiedIndexMutations,
+            Map<TableInfo, List<Mutation>> verifiedOrDeletedIndexMutations) throws SQLException {
+        Iterator<Entry<TableInfo, List<Mutation>>> mapIter = mutationMap.entrySet().iterator();
+        while (mapIter.hasNext()) {
+            Entry<TableInfo, List<Mutation>> pair = mapIter.next();
+            TableInfo tableInfo = pair.getKey();
+            if (IndexUtil.isGlobalIndexCheckerEnabled(connection, tableInfo.getHTableName())) {
+                PTable table = tableInfo.getOrigTableRef().getTable();
+                byte[] emptyCF = SchemaUtil.getEmptyColumnFamily(table);
+                byte[] emptyCQ = EncodedColumnsUtil.getEmptyKeyValueInfo(table).getFirst();
+
+                List<Mutation> mutations = pair.getValue();
+
+                if (mutations.size() == 0) {
+                    continue;
+                }
+
+                for (Mutation m : mutations) {
+                    if (m instanceof Delete) {
+                        Put put = new Put(m.getRow());
+                        put.addColumn(emptyCF, emptyCQ, m.getTimestamp(), IndexRegionObserver.UNVERIFIED_BYTES);
+                        // The Delete gets marked as unverified in Phase 1 and gets deleted on Phase 3.
+                        addToMap(unverifiedIndexMutations, tableInfo, put);
+                        addToMap(verifiedOrDeletedIndexMutations, tableInfo, m);
+                    } else {
+                        // These mutations already have Unverified_Bytes set.
+                        addToMap(unverifiedIndexMutations, tableInfo, m);
+                        // We will need to have verified byte set version for the Phase 3.
```

Comment:
Nit: "We will need .." sounds like that is being done somewhere else. Perhaps rephrase as "set verified bytes for phase3"

Reply:
ok